Include bonds in communication cutoff estimate

akohlmey opened the pull request:
**Summary**

Add a function to the `Comm` class that determines the communication cutoff consistently. When determining the communication cutoff use the larger of `Neighbor::maxneighcut` and the user provided communication cutoff, but also consider bond lengths and whether angles and dihedrals are present. If there is no pair style and no user provided cutoff, use the bond length inferred value, otherwise print a warning, if that estimate is larger than the actual communication cutoff. Also print a warning message, when the user specified communication cutoff is overridden.

**Related Issues**

Fixes #1577 

**Author(s)**

Axel Kohlmeyer (Temple U)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

yes.

**Post Submission Checklist**
- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system

sjplimp commented:
A couple of problems I see.  First, the neigh cutoff makes a big difference in performance, so we don't want to set it larger than the user expects, esp in a hidden manner.  If I follow the logic, it seems like it is setting if to the larger of the pair cutoff and 3.5*maxbond (if dihedrals exist).  Which could be much larger than a particular model needs.  We already have a good error message if a dihedral atom is not found.  Second, the comm modify cutoff setting is added to these values.  So the user may have already set that value appropriate for their model, but this change makes the cutoff even longer.

So I suggest we only set the cutoff to 1.5*maxbondlen if no pair exists *and* no comm modify cufoff settting exists.  And if we do that we issue a warning that the cutoff was guesstimated for those 2 reasons.  Otherwise do nothing.  We could also flag warnings (in comm) if the final cutoff is less than 1.5x (or 3.5x for dihedrals) than maxbondlen.

akohlmey commented:
> A couple of problems I see. First, the neigh cutoff makes a big difference in performance, so we don't want to set it larger than the user expects, esp in a hidden manner. If I follow the logic, it seems like it is setting if to the larger of the pair cutoff and 3.5*maxbond (if dihedrals exist). Which could be much larger than a particular model needs. We already have a good error message if a dihedral atom is not found. 

in a typical molecular system, equilibrium bond lengths are between 1.0 and 1.7 \AA, so 3.5 * 1.7 would be 5.95 and thus much shorter than typical non-bonded cutoffs. 

>Second, the comm modify cutoff setting is added to these values. So the user may have already set that value appropriate for their model, but this change makes the cutoff even longer.

that is not correct. the communication cutoff is set to the larger of either `comm_modify cutoff` or the maximum neighbor cutoff.
 
> So I suggest we only set the cutoff to 1.5*maxbondlen if no pair exists _and_ no comm modify cufoff settting exists. And if we do that we issue a warning that the cutoff was guesstimated for those 2 reasons. Otherwise do nothing. We could also flag warnings (in comm) if the final cutoff is less than 1.5x (or 3.5x for dihedrals) than maxbondlen.

we do a lot of things behind people's backs.  i give you, that i may have been too careful. if i consider that we store angles with the central atom and dihedrals/impropers with the second atom, we don't need to consider *all* bonds. so perhaps 1.5x for bonds and angles, and 2.0x for dihedrals/impropers might be a compromise that should still cover the vast majority of problematic cases and significantly reduce the number of cases where there would be a performance impact (and the magnitude of it) when this is overestimating the cutoff.

sjplimp commented:
>that is not correct. the communication cutoff is set to the larger of either comm_modify cutoff or the >maximum neighbor cutoff.

you're right on this - I mis-read the current doc page.

But if the user sets a non-zero value, I don't think we should override it with the bond calculation.
However we should continue to override it with the pair+skin value.   Since the code is guaranteed
to do the wrong thing silently if we don't - i.e. the pair and comm modify cutoff settings are inconsistent.
We should probably add a warning that we did that.

But if pair and/or comm_modify are set, I say leave the cutoff alone.  But we can add a warning that the final value < 1.5*bond or 2*bond (for dihedral) - i.e. your new limits.

If no pair and no comm modify are set, then auto-set the cutoff via the maxbond (1.5x or 2x)
But also issue a warning to tell them it was done for them, so they realize it.

akohlmey commented:
> > that is not correct. the communication cutoff is set to the larger of either comm_modify cutoff or the >maximum neighbor cutoff.
> 
> you're right on this - I mis-read the current doc page.
> 
> But if the user sets a non-zero value, I don't think we should override it with the bond calculation.
> However we should continue to override it with the pair+skin value. Since the code is guaranteed
> to do the wrong thing silently if we don't - i.e. the pair and comm modify cutoff settings are inconsistent.
> We should probably add a warning that we did that.

if we are going to override, we should do it consistently. either for all or for none. whether we take the neighbor list cutoff or a bond length derived cutoff, it should not be handled differently. what makes one a better heuristic than the other? 

i've revised to code and moved the bond length based estimate out of the neighbor class. putting it there was mostly laziness. since this is only really needed for communication cutoff, it is now a function in the Comm class. so no more negative impact on the neighborlist maximum cutoff and neighbor list build performance is possible.

> But if pair and/or comm_modify are set, I say leave the cutoff alone. But we can add a warning that the final value < 1.5_bond or 2_bond (for dihedral) - i.e. your new limits.
> 
> If no pair and no comm modify are set, then auto-set the cutoff via the maxbond (1.5x or 2x)
> But also issue a warning to tell them it was done for them, so they realize it.

ok. i've added a warning message to the getter function that will print a warning when the explicitly specified communication cutoff is increased. regardless of the heuristic.

sjplimp commented:
>if we are going to override, we should do it consistently. either for all or for none. whether we take the >neighbor list cutoff or a bond length derived cutoff, it should not be handled differently. what makes >one a better heuristic than the other?

B/c the pair cutoff + skin is not a heuristic - it's the precise required comm cutoff for the pair style to work, as set by the user, when they specified the pair style.  The bond-length cutoff is a (decent) guesstimate that could be just right or too big or too small for a particular system.  If the user used comm_modify cutoff then I think we should trust them and do what they ask, but warn them if it's smaller than what the heuristic thinks might be needed.

sjplimp commented:
Sorry, one more iteration.  This is what I said earlier:

>But if pair and/or comm_modify are set, I say leave the cutoff alone. But we can add a warning that >the final value < 1.5bond or 2bond (for dihedral) - i.e. your new limits.

I believe what the latest is doing is overriding the pair cutoff if the bond cutoff is longer.  I think
we should not do that, but rather add a warning that the pair cutoff may be insufficient for bond finding, so use comm_modify cutoff if necessary.

Also I realized that for newton on, 1.5 and 2 are good.  But for newton off, the guesstimate
should be up to ~3x (for dihedrals), i.e. a 1-4 distance.

akohlmey commented:
@sjplimp while I don't agree 100% with your reasoning, I don't consider this important enough to argue it now and instead have tried to implement what you are asking for. We can talk about this and some other default choices in LAMMPS better in person.

akohlmey commented:
@sjplimp is this now to your liking?
